src/nodebuilder.cpp

```diff
@@ -69,11 +71,38 @@ void NodeBuilder::OnMapStart(const Mark& mark, const std::string& tag,
   node.set_tag(tag);
   node.set_style(style);
   m_mapDepth++;
+  m_mergeDicts.emplace_back();
+}
+
+void MergeMapCollection(detail::node& map_to, detail::node& map_from,
+                        detail::shared_memory_holder& pMemory) {
+  for (auto j = map_from.begin(); j != map_from.end(); j++) {
+    const auto from_key = j->first;
+    /// NOTE: const_map_to.get(*j->first) cannot be used here, since it
+    /// compares only the shared_ptr's, while we need to compare the key
+    /// itself.
+    ///
+    /// NOTE: get() also iterates over elements
+    bool found = std::any_of(map_to.begin(), map_to.end(), [&](const detail::node_iterator_value<detail::node> & kv)
+    {
+        const auto key_node = kv.first;
+        return key_node->scalar() == from_key->scalar();
+    });
+    if (!found)
+      map_to.insert(*from_key, *j->second, pMemory);
+  }
 }
 
 void NodeBuilder::OnMapEnd() {
   assert(m_mapDepth > 0);
+  detail::node& collection = *m_stack.back();
+  auto& toMerge = *m_mergeDicts.rbegin();
+  /// The elements for merging should be traversed in reverse order to prefer last values.
+  for (auto it = toMerge.rbegin(); it != toMerge.rend(); ++it) {
+    MergeMapCollection(collection, **it, m_pMemory);
```

Comment:
If `MergeMapCollection` is being adjusted as suggested above `**it` must be changed to `std::move(**it)`